### PR TITLE
Use Toolchains and Java 17

### PR DESCRIPTION
**Pull request by leonard84**

### CODE REVIEW

File: .github/workflows/branches-and-prs.yml [hunk 1]
```diff
@@ -45,21 +45,35 @@ jobs:
...
+        shell: bash
+        run: echo "JDK${{ matrix.java }}=$JAVA_HOME" >> $GITHUB_ENV
+      - name: 'Set up JDK 8'
+        uses: actions/setup-java@v2
+        with:
+          java-version: 8
+          distribution: 'adopt'
+      - name: Prepare JDK8 env var
+        shell: bash
+        run: echo "JDK8=$JAVA_HOME" >> $GITHUB_ENV
```
Comment (szpak) [inline .github/workflows/branches-and-prs.yml:64]: I haven't played with Java 16 on GH Actions before, so probably I miss something, but why you have separate blocks for Java 8 and others?

File: .github/workflows/branches-and-prs.yml [hunk 2]
```diff
@@ -17,7 +17,7 @@ jobs:
       matrix:
         os: ['ubuntu-latest']
         variant: ['2.5', '3.0']
-        java: ['8', '11', '15']
+        java: ['8', '11', '16']
```
Comment (szpak) [inline .github/workflows/branches-and-prs.yml:20]: Groovy 2.5 doesn't seem to support Java 16. You should probably exclude those builds.

File: .github/workflows/branches-and-prs.yml (hunk 1 above)
Comment (leonard84) [reply .github/workflows/branches-and-prs.yml:64]: WDYM? With toolchains we can now properly build the `main` sources with JDK 8, while executing the tests with whatever JDK we want, which matches the real world much more closely. To make the JDKs properly discoverable we have to put them in env vars so gradle can pick them up.

File: .github/workflows/branches-and-prs.yml (hunk 2 above)
Comment (kriegaex) [reply .github/workflows/branches-and-prs.yml:20]: I would not exclude, but configure to compile to a lower target. Running the build on a recent JDK is still important, IMO.

File: .github/workflows/branches-and-prs.yml (hunk 1 above)
Comment (szpak) [reply .github/workflows/branches-and-prs.yml:64]: Ok, I missed that you setup two JDK versions 8 and X which was problematic without toolchains. That might make sense as is closer to reality (having released Spock built with JDK 8).

File: .github/workflows/branches-and-prs.yml (hunk 2 above)
Comment (szpak) [reply .github/workflows/branches-and-prs.yml:20]: You might be right. Groovy 2.5 might be able to be executed with Java 16, but with lower target version (but I haven't tested it).

Comment (kriegaex) [reply .github/workflows/branches-and-prs.yml:20]: I upgraded my Spock playground project, and indeed it is working, despite a few compiler quirks in Groovy-Eclipse which forced me to refactor two classes. Also IntelliJ IDEA is bitching around now. But I got it compiling in Maven. Please also refer to
  * https://github.com/groovy/groovy-eclipse/issues/1292
  * https://youtrack.jetbrains.com/issue/IDEA-275889

But in general, it should work.

**Comment by leonard84** [comment]
Seems like Java 17 is only supported by Groovy 4.x

**Comment by szpak** [comment]
> Seems like Java 17 is only supported by Groovy 4.x

Do you mean by a runtime JDK? I was building today one of my projects with Gradle 7.2, OpenJDK 17 and Spock 2.0 with Groovy 3.0.9 (in fact with 3.0.8 I had an error with `Unssuported...`).

File: gradle.properties [hunk 3]
```diff
@@ -14,7 +14,8 @@
 # limitations under the License.
 #
 
-
+org.gradle.java.installations.auto-download=false
+org.gradle.java.installations.fromEnv=JDK8,JDK11,JDK16,JDK17
```
Comment (szpak) [inline gradle.properties:18]:
```suggestion
org.gradle.java.installations.fromEnv=JDK8,JDK11,JDK17
```
?

File: gradle.properties [hunk 4]
```diff
@@ -14,7 +14,8 @@
...
```
Comment (szpak) [inline gradle.properties:19]: Related to the above - as JDKx is set by you, I wonder, if:
> org.gradle.java.installations.auto-detect=false

couldn't be added to the configuration?

I see that in the 3.0, 11 build on Ubuntu, JDK 11 is "Detected by: Common Linux Locations", but probably it just has a precedence, as the JDK11 variable also seems to be set.
https://github.com/spockframework/spock/pull/1366/checks?check_run_id=3748208999#step:7:55

**Review by szpak — CHANGES_REQUESTED** [review]

File: gradle.properties (hunk 4 above)
Comment (leonard84) [reply gradle.properties:19]: This would force everyone to define those env variables on their workstations as well.

Comment (szpak) [reply gradle.properties:19]: In that case, `org.gradle.java.installations.auto-detect=false` in general could be problematic. We could add it only on CI (via `echo "..." >> gradle.properties`), but assuming the locally pre-installed (and auto-detected) version is ok for us (not that one pointed out with the env) there should be no problem and the gain of avoiding auto-detection would be probably negligible.

**Review by szpak — APPROVED** [review]

**Pull request merged**